In [353]:
import pandas as pd
import os

# Lấy dữ liệu

In [354]:

pathfile = 'D:\Work\MindX\BI\BI40'
df = pd.read_csv(os.path.join(pathfile,"netflix_titles.csv"))
df

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

## Check thông tin DF

- Dữ liệu có 8807 dòng tương ứng với mỗi cột sẽ cần có 8807 dòng giá trị

- Sau khi kiểm tra dữ liệu những cột xuất hiện thiếu dữ liệu bao gồm:
    
    + director

    + cast

    + country

    + date_added

    + rating

    + duration

- Quá trình làm sạch dữ liệu bao gồm:

    + Kiểm tra và xử lý dữ liệu null

    + Kiểm tra và chuyển đổi kiểu dữ liệu tương ứng với từng cột dữ liệu

In [355]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


## Kiểm tra và xử lý dữ liệu null 

- Ta tiến hành xử lý dữ liệu từ cột có số lượng giá trị null ít nhất

### duration

- 3 giá trị null có giá trị bị lệch cột

- giá trị của cột đan nằm trong cột rating

- ta sẽ thay thế null = giá trị trong cột rating

- giá trị tại cột rating sẽ thay thế bằng giá trị orther

In [356]:
df[df['duration'].isnull()]

show_id   type                                 title    director  \
5541   s5542  Movie                       Louis C.K. 2017  Louis C.K.   
5794   s5795  Movie                 Louis C.K.: Hilarious  Louis C.K.   
5813   s5814  Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.   

            cast        country          date_added  release_year  rating  \
5541  Louis C.K.  United States       April 4, 2017          2017  74 min   
5794  Louis C.K.  United States  September 16, 2016          2010  84 min   
5813  Louis C.K.  United States     August 15, 2016          2015  66 min   

     duration listed_in                                        description  
5541      NaN    Movies  Louis C.K. muses on religion, eternal love, gi...  
5794      NaN    Movies  Emmy-winning comedy writer Louis C.K. brings h...  
5813      NaN    Movies  The comic puts his trademark hilarious/thought...

In [357]:
df['duration'] = df['rating'].where(df['duration'].isnull()==True,df['duration'])
df[df['show_id'].isin(['s5542','s5795','s5814'])]

show_id   type                                 title    director  \
5541   s5542  Movie                       Louis C.K. 2017  Louis C.K.   
5794   s5795  Movie                 Louis C.K.: Hilarious  Louis C.K.   
5813   s5814  Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.   

            cast        country          date_added  release_year  rating  \
5541  Louis C.K.  United States       April 4, 2017          2017  74 min   
5794  Louis C.K.  United States  September 16, 2016          2010  84 min   
5813  Louis C.K.  United States     August 15, 2016          2015  66 min   

     duration listed_in                                        description  
5541   74 min    Movies  Louis C.K. muses on religion, eternal love, gi...  
5794   84 min    Movies  Emmy-winning comedy writer Louis C.K. brings h...  
5813   66 min    Movies  The comic puts his trademark hilarious/thought...

In [358]:
df['rating'] = df['rating'].apply(lambda x: 'Other' if 'min' in str(x) else x)
df[df['show_id'].isin(['s5542','s5795','s5814'])]

show_id   type                                 title    director  \
5541   s5542  Movie                       Louis C.K. 2017  Louis C.K.   
5794   s5795  Movie                 Louis C.K.: Hilarious  Louis C.K.   
5813   s5814  Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.   

            cast        country          date_added  release_year rating  \
5541  Louis C.K.  United States       April 4, 2017          2017  Other   
5794  Louis C.K.  United States  September 16, 2016          2010  Other   
5813  Louis C.K.  United States     August 15, 2016          2015  Other   

     duration listed_in                                        description  
5541   74 min    Movies  Louis C.K. muses on religion, eternal love, gi...  
5794   84 min    Movies  Emmy-winning comedy writer Louis C.K. brings h...  
5813   66 min    Movies  The comic puts his trademark hilarious/thought...

In [359]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8807 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


### rating 

- Thay giá trị null bằng 'Other'

In [360]:
df[df['rating'].isnull()]

show_id     type                                              title  \
5989   s5990    Movie  13TH: A Conversation with Oprah Winfrey & Ava ...   
6827   s6828  TV Show                  Gargantia on the Verdurous Planet   
7312   s7313  TV Show                                       Little Lunch   
7537   s7538    Movie                               My Honor Was Loyalty   

             director                                               cast  \
5989              NaN                        Oprah Winfrey, Ava DuVernay   
6827              NaN  Kaito Ishikawa, Hisako Kanemoto, Ai Kayano, Ka...   
7312              NaN  Flynn Curry, Olivia Deeble, Madison Lu, Oisín ...   
7537  Alessandro Pepe  Leone Frisa, Paolo Vaccarino, Francesco Miglio...   

        country        date_added  release_year rating  duration  \
5989        NaN  January 26, 2017          2017    NaN    37 min   
6827      Japan  December 1, 2016          2013    NaN  1 Season   
7312  Australia  February 1, 2018          2015    NaN  1 Season   
7537      Italy     March 1, 2017          2015    NaN   115 min   

                                 listed_in  \
5989                                Movies   
6827  Anime Series, International TV Shows   
7312                 Kids' TV, TV Comedies   
7537                                Dramas   

                                            description  
5989  Oprah Winfrey sits down with director Ava DuVe...  
6827  After falling through a wormhole, a space-dwel...  
7312  Adopting a child's perspective, this show take...  
7537  Amid the chaos and horror of World War II, a c...

In [361]:
list(set(df['rating']))

[nan,
 'TV-Y7-FV',
 'PG-13',
 'PG',
 'TV-14',
 'NC-17',
 'TV-Y',
 'TV-Y7',
 'G',
 'TV-G',
 'Other',
 'NR',
 'R',
 'TV-PG',
 'TV-MA',
 'UR']

In [362]:
df['rating'].fillna(value='Other',inplace=True)

In [363]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8807 non-null   object
 9   duration      8807 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


### date_added

- là ngày phát hành => không có khả năng fill bằng value nằm trong cột hoặc sẻach rồi thay thế

- Phương án:

    - do cột release_year vẫn còn dữ liệu năm => có thể thay thế bằng ngày 01/01/release_year => January 1, release_year

In [364]:
llist_check_null = ['s7407', 's8183', 's6067', 's6807', 's6796', 's7848', 's6902', 's7255', 's7197', 's6175']

In [365]:
df[df['date_added'].isnull()]

show_id     type                                        title director  \
6066   s6067  TV Show  A Young Doctor's Notebook and Other Stories      NaN   
6174   s6175  TV Show              Anthony Bourdain: Parts Unknown      NaN   
6795   s6796  TV Show                                      Frasier      NaN   
6806   s6807  TV Show                                      Friends      NaN   
6901   s6902  TV Show                              Gunslinger Girl      NaN   
7196   s7197  TV Show                                     Kikoriki      NaN   
7254   s7255  TV Show                          La Familia P. Luche      NaN   
7406   s7407  TV Show                                        Maron      NaN   
7847   s7848  TV Show                                 Red vs. Blue      NaN   
8182   s8183  TV Show                 The Adventures of Figaro Pho      NaN   

                                                   cast         country  \
6066  Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...  United Kingdom   
6174                                   Anthony Bourdain   United States   
6795  Kelsey Grammer, Jane Leeves, David Hyde Pierce...   United States   
6806  Jennifer Aniston, Courteney Cox, Lisa Kudrow, ...   United States   
6901  Yuuka Nanri, Kanako Mitsuhashi, Eri Sendai, Am...           Japan   
7196                                      Igor Dmitriev             NaN   
7254  Eugenio Derbez, Consuelo Duval, Luis Manuel Áv...   United States   
7406  Marc Maron, Judd Hirsch, Josh Brener, Nora Zeh...   United States   
7847  Burnie Burns, Jason Saldaña, Gustavo Sorola, G...   United States   
8182  Luke Jurevicius, Craig Behenna, Charlotte Haml...       Australia   

     date_added  release_year rating    duration  \
6066        NaN          2013  TV-MA   2 Seasons   
6174        NaN          2018  TV-PG   5 Seasons   
6795        NaN          2003  TV-PG  11 Seasons   
6806        NaN          2003  TV-14  10 Seasons   
6901        NaN          2008  TV-14   2 Seasons   
7196        NaN          2010   TV-Y   2 Seasons   
7254        NaN          2012  TV-14   3 Seasons   
7406        NaN          2016  TV-MA   4 Seasons   
7847        NaN          2015     NR  13 Seasons   
8182        NaN          2015  TV-Y7   2 Seasons   

                                              listed_in  \
6066           British TV Shows, TV Comedies, TV Dramas   
6174                                         Docuseries   
6795                     Classic & Cult TV, TV Comedies   
6806                     Classic & Cult TV, TV Comedies   
6901                       Anime Series, Crime TV Shows   
7196                                           Kids' TV   
7254  International TV Shows, Spanish-Language TV Sh...   
7406                                        TV Comedies   
7847  TV Action & Adventure, TV Comedies, TV Sci-Fi ...   
8182                              Kids' TV, TV Comedies   

                                            description  
6066  Set during the Russian Revolution, this comic ...  
6174  This CNN original series has chef Anthony Bour...  
6795  Frasier Crane is a snooty but lovable Seattle ...  
6806  This hit sitcom follows the merry misadventure...  
6901  On the surface, the Social Welfare Agency appe...  
7196  A wacky rabbit and his gang of animal pals hav...  
7254  This irreverent sitcom featues Ludovico, Feder...  
7406  Marc Maron stars as Marc Maron, who interviews...  
7847  This parody of first-person shooter games, mil...  
8182  Imagine your worst fears, then multiply them: ...

In [366]:
df['date_added'].fillna(value='January 1, ', inplace=True)
df[df['date_added']=='January 1, ']
df["date_added"] = df["date_added"].str.cat(df["release_year"].astype('str')).where(df['date_added']=='January 1, ',df['date_added'])
df[df['show_id'].isin(llist_check_null)]

show_id     type                                        title director  \
6066   s6067  TV Show  A Young Doctor's Notebook and Other Stories      NaN   
6174   s6175  TV Show              Anthony Bourdain: Parts Unknown      NaN   
6795   s6796  TV Show                                      Frasier      NaN   
6806   s6807  TV Show                                      Friends      NaN   
6901   s6902  TV Show                              Gunslinger Girl      NaN   
7196   s7197  TV Show                                     Kikoriki      NaN   
7254   s7255  TV Show                          La Familia P. Luche      NaN   
7406   s7407  TV Show                                        Maron      NaN   
7847   s7848  TV Show                                 Red vs. Blue      NaN   
8182   s8183  TV Show                 The Adventures of Figaro Pho      NaN   

                                                   cast         country  \
6066  Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...  United Kingdom   
6174                                   Anthony Bourdain   United States   
6795  Kelsey Grammer, Jane Leeves, David Hyde Pierce...   United States   
6806  Jennifer Aniston, Courteney Cox, Lisa Kudrow, ...   United States   
6901  Yuuka Nanri, Kanako Mitsuhashi, Eri Sendai, Am...           Japan   
7196                                      Igor Dmitriev             NaN   
7254  Eugenio Derbez, Consuelo Duval, Luis Manuel Áv...   United States   
7406  Marc Maron, Judd Hirsch, Josh Brener, Nora Zeh...   United States   
7847  Burnie Burns, Jason Saldaña, Gustavo Sorola, G...   United States   
8182  Luke Jurevicius, Craig Behenna, Charlotte Haml...       Australia   

           date_added  release_year rating    duration  \
6066  January 1, 2013          2013  TV-MA   2 Seasons   
6174  January 1, 2018          2018  TV-PG   5 Seasons   
6795  January 1, 2003          2003  TV-PG  11 Seasons   
6806  January 1, 2003          2003  TV-14  10 Seasons   
6901  January 1, 2008          2008  TV-14   2 Seasons   
7196  January 1, 2010          2010   TV-Y   2 Seasons   
7254  January 1, 2012          2012  TV-14   3 Seasons   
7406  January 1, 2016          2016  TV-MA   4 Seasons   
7847  January 1, 2015          2015     NR  13 Seasons   
8182  January 1, 2015          2015  TV-Y7   2 Seasons   

                                              listed_in  \
6066           British TV Shows, TV Comedies, TV Dramas   
6174                                         Docuseries   
6795                     Classic & Cult TV, TV Comedies   
6806                     Classic & Cult TV, TV Comedies   
6901                       Anime Series, Crime TV Shows   
7196                                           Kids' TV   
7254  International TV Shows, Spanish-Language TV Sh...   
7406                                        TV Comedies   
7847  TV Action & Adventure, TV Comedies, TV Sci-Fi ...   
8182                              Kids' TV, TV Comedies   

                                            description  
6066  Set during the Russian Revolution, this comic ...  
6174  This CNN original series has chef Anthony Bour...  
6795  Frasier Crane is a snooty but lovable Seattle ...  
6806  This hit sitcom follows the merry misadventure...  
6901  On the surface, the Social Welfare Agency appe...  
7196  A wacky rabbit and his gang of animal pals hav...  
7254  This irreverent sitcom featues Ludovico, Feder...  
7406  Marc Maron stars as Marc Maron, who interviews...  
7847  This parody of first-person shooter games, mil...  
8182  Imagine your worst fears, then multiply them: ...

In [367]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8807 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8807 non-null   object
 9   duration      8807 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


### director, cast

- Dữ liệu này không thể tìm kiếm các cột khác tương ứng lại chứa dữ liệu => thay thế bằng giá trị unknown


In [368]:
df['director'].fillna(value='unknow',inplace=True)
df['cast'].fillna(value='unknow',inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      8807 non-null   object
 4   cast          8807 non-null   object
 5   country       7976 non-null   object
 6   date_added    8807 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8807 non-null   object
 9   duration      8807 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


### country

- Dữ liệu country có thể làm theo kiểu đơn gian là ta cũng thay dữ liệu null bằng giá trị unknow do lượng dữ liệu thiếu khá lớn

- Tuy nhiên trong trường hợp cần cứu dữ liệu tối đa có thể ta có thể chia dữ liệu null làm 2 hoặc 3 loại dữ liệu ở đây mình sẽ chia làm 2 loại:

    + Loại 1: dữ liêu null ở cả director và country => với dữ liệu này ta chỉ có thể thay bằng unknow

    + Loại 2: dữ liệu chỉ null ở country và có dữ liệu director, tại đây ta sẽ thay dữ liệu null bằng dữ liệu country của một phim khác có cùng director
    
      độ chính xác tầm 80%


### Loại 1:

In [369]:
df[(df['country'].isnull())&(df['director']=='unknow')]

show_id     type                                title director  \
3         s4  TV Show                Jailbirds New Orleans   unknow   
10       s11  TV Show  Vendetta: Truth, Lies and The Mafia   unknow   
14       s15  TV Show      Crime Stories: India Detectives   unknow   
19       s20  TV Show                               Jaguar   unknow   
31       s32  TV Show                   Chicago Party Aunt   unknow   
...      ...      ...                                  ...      ...   
8679   s8680  TV Show                   ViR: The Robot Boy   unknow   
8690   s8691  TV Show                              Wake Up   unknow   
8783   s8784  TV Show                                 Yoko   unknow   
8785   s8786  TV Show                                  YOM   unknow   
8803   s8804  TV Show                          Zombie Dumb   unknow   

                                                   cast country  \
3                                                unknow     NaN   
10                                               unknow     NaN   
14                                               unknow     NaN   
19    Blanca Suárez, Iván Marcos, Óscar Casas, Adriá...     NaN   
31    Lauren Ash, Rory O'Malley, RuPaul Charles, Jil...     NaN   
...                                                 ...     ...   
8679  Vidit Kumar, Roopa Bhimani, Anubhav Saha, Bria...     NaN   
8690  Jag Huang, Chris Wu, Tiffany Hsu, Michael Huan...     NaN   
8783  Eileen Stevens, Alyson Leigh Rosenfeld, Sarah ...     NaN   
8785  Sairaj, Devyani Dagaonkar, Ketan Singh, Mayur ...     NaN   
8803                                             unknow     NaN   

              date_added  release_year rating   duration  \
3     September 24, 2021          2021  TV-MA   1 Season   
10    September 24, 2021          2021  TV-MA   1 Season   
14    September 22, 2021          2021  TV-MA   1 Season   
19    September 22, 2021          2021  TV-MA   1 Season   
31    September 17, 2021          2021  TV-MA   1 Season   
...                  ...           ...    ...        ...   
8679      March 31, 2018          2013  TV-Y7  2 Seasons   
8690      March 31, 2018          2017  TV-14  2 Seasons   
8783       June 23, 2018          2016   TV-Y   1 Season   
8785        June 7, 2018          2016  TV-Y7   1 Season   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   

                                              listed_in  \
3                                Docuseries, Reality TV   
10    Crime TV Shows, Docuseries, International TV S...   
14         British TV Shows, Crime TV Shows, Docuseries   
19    International TV Shows, Spanish-Language TV Sh...   
31                                          TV Comedies   
...                                                 ...   
8679                                           Kids' TV   
8690                  International TV Shows, TV Dramas   
8783                                           Kids' TV   
8785                                           Kids' TV   
8803             Kids' TV, Korean TV Shows, TV Comedies   

                                            description  
3     Feuds, flirtations and toilet talk go down amo...  
10    Sicily boasts a bold "Anti-Mafia" coalition. B...  
14    Cameras following Bengaluru police on the job ...  
19    In the 1960s, a Holocaust survivor joins a gro...  
31    Chicago Party Aunt Diane is an idolized troubl...  
...                                                 ...  
8679  Humanoid robot boy ViR lives with his inventor...  
8690  After taking the blame for a patient death, an...  
8783  Friends Mai, Oto and Vik's games at the park b...  
8785  With the mind of a human being, and the body o...  
8803  While living alone in a spooky town, a young g...  

[409 rows x 12 columns]

In [370]:
df['country'] = df['director'].where(df['director']=='unknow',df['country'])

In [371]:
df[df['country']=='unknow']

show_id     type                                title director  \
1         s2  TV Show                        Blood & Water   unknow   
3         s4  TV Show                Jailbirds New Orleans   unknow   
4         s5  TV Show                         Kota Factory   unknow   
10       s11  TV Show  Vendetta: Truth, Lies and The Mafia   unknow   
14       s15  TV Show      Crime Stories: India Detectives   unknow   
...      ...      ...                                  ...      ...   
8795   s8796  TV Show                      Yu-Gi-Oh! Arc-V   unknow   
8796   s8797  TV Show                           Yunus Emre   unknow   
8797   s8798  TV Show                            Zak Storm   unknow   
8800   s8801  TV Show                   Zindagi Gulzar Hai   unknow   
8803   s8804  TV Show                          Zombie Dumb   unknow   

                                                   cast country  \
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...  unknow   
3                                                unknow  unknow   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...  unknow   
10                                               unknow  unknow   
14                                               unknow  unknow   
...                                                 ...     ...   
8795  Mike Liscio, Emily Bauer, Billy Bob Thompson, ...  unknow   
8796  Gökhan Atalay, Payidar Tüfekçioglu, Baran Akbu...  unknow   
8797  Michael Johnston, Jessica Gee-George, Christin...  unknow   
8800  Sanam Saeed, Fawad Khan, Ayesha Omer, Mehreen ...  unknow   
8803                                             unknow  unknow   

              date_added  release_year rating   duration  \
1     September 24, 2021          2021  TV-MA  2 Seasons   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
10    September 24, 2021          2021  TV-MA   1 Season   
14    September 22, 2021          2021  TV-MA   1 Season   
...                  ...           ...    ...        ...   
8795         May 1, 2018          2015  TV-Y7  2 Seasons   
8796    January 17, 2017          2016  TV-PG  2 Seasons   
8797  September 13, 2018          2016  TV-Y7  3 Seasons   
8800   December 15, 2016          2012  TV-PG   1 Season   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   

                                              listed_in  \
1       International TV Shows, TV Dramas, TV Mysteries   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
10    Crime TV Shows, Docuseries, International TV S...   
14         British TV Shows, Crime TV Shows, Docuseries   
...                                                 ...   
8795                             Anime Series, Kids' TV   
8796                  International TV Shows, TV Dramas   
8797                                           Kids' TV   
8800  International TV Shows, Romantic TV Shows, TV ...   
8803             Kids' TV, Korean TV Shows, TV Comedies   

                                            description  
1     After crossing paths at a party, a Cape Town t...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
10    Sicily boasts a bold "Anti-Mafia" coalition. B...  
14    Cameras following Bengaluru police on the job ...  
...                                                 ...  
8795  Now that he's discovered the Pendulum Summonin...  
8796  During the Mongol invasions, Yunus Emre leaves...  
8797  Teen surfer Zak Storm is mysteriously transpor...  
8800  Strong-willed, middle-class Kashaf and carefre...  
8803  While living alone in a spooky town, a young g...  

[2634 rows x 12 columns]

### Loại 2:

In [372]:
df[(df['country'].isnull())]

show_id     type                             title  \
2         s3  TV Show                         Ganglands   
5         s6  TV Show                     Midnight Mass   
6         s7    Movie  My Little Pony: A New Generation   
11       s12  TV Show                  Bangkok Breaking   
13       s14    Movie  Confessions of an Invisible Girl   
...      ...      ...                               ...   
8585   s8586    Movie             Three-Quarters Decent   
8602   s8603    Movie     Tom and Jerry: The Magic Ring   
8622   s8623    Movie            Tremors 2: Aftershocks   
8718   s8719    Movie            Westside vs. the World   
8759   s8760    Movie            World's Weirdest Homes   

                           director  \
2                   Julien Leclercq   
5                     Mike Flanagan   
6     Robert Cullen, José Luis Ucha   
11                Kongkiat Komesiri   
13                    Bruno Garotti   
...                             ...   
8585                  Mohamed Hamdy   
8602                     Phil Roman   
8622                    S.S. Wilson   
8718                  Michael Fahey   
8759               Storm Theunissen   

                                                   cast country  \
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...     NaN   
5     Kate Siegel, Zach Gilford, Hamish Linklater, H...     NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...     NaN   
11    Sukollawat Kanarot, Sushar Manaying, Pavarit M...     NaN   
13    Klara Castanho, Lucca Picon, Júlia Gomes, Marc...     NaN   
...                                                 ...     ...   
8585  Mohamed Ragab, Lamitta Frangieh, Mohsen Mansou...     NaN   
8602  Richard Kind, Dana Hill, Anndi McAfee, Tony Ja...     NaN   
8622  Fred Ward, Chris Gartin, Helen Shaver, Michael...     NaN   
8718                         Ron Perlman, Louie Simmons     NaN   
8759                                     Charlie Luxton     NaN   

              date_added  release_year rating  duration  \
2     September 24, 2021          2021  TV-MA  1 Season   
5     September 24, 2021          2021  TV-MA  1 Season   
6     September 24, 2021          2021     PG    91 min   
11    September 23, 2021          2021  TV-MA  1 Season   
13    September 22, 2021          2021  TV-PG    91 min   
...                  ...           ...    ...       ...   
8585       June 20, 2019          2010  TV-14    96 min   
8602   December 15, 2019          2001  TV-Y7    60 min   
8622     January 1, 2020          1995  PG-13   100 min   
8718      August 9, 2019          2019  TV-MA    96 min   
8759    February 1, 2019          2015  TV-PG    49 min   

                                              listed_in  \
2     Crime TV Shows, International TV Shows, TV Act...   
5                    TV Dramas, TV Horror, TV Mysteries   
6                              Children & Family Movies   
11    Crime TV Shows, International TV Shows, TV Act...   
13                   Children & Family Movies, Comedies   
...                                                 ...   
8585             Comedies, Dramas, International Movies   
8602                 Children & Family Movies, Comedies   
8622          Comedies, Horror Movies, Sci-Fi & Fantasy   
8718                       Documentaries, Sports Movies   
8759                                             Movies   

                                            description  
2     To protect his family from a powerful drug lor...  
5     The arrival of a charismatic young priest brin...  
6     Equestria's divided. But a bright-eyed hero be...  
11    Struggling to earn a living in Bangkok, a man ...  
13    When the clever but socially-awkward Tetê join...  
...                                                 ...  
8585  Determined to fight corruption in his country,...  
8602  When a young wizard leaves Tom to guard his pr...  
8622  A rag-tag team of survivalists and scientists ...  
8718  A look into the jo

In [373]:
# lấy ra list danh sách director cần sửa
directorlist = list(set(df[(df['country'].isnull())]['director']))

In [377]:
df[(df['director']=='Julien Leclercq') & (df['country'].isnull())==False]['country'].iloc[0]

'United States'

In [374]:
for i in directorlist:
    a = df[(df['director']==i) & (df['country'].isnull())==False]['country'].iloc[0] # Lấy ra giá trị country tương ứng với director
    df['country'] = df.apply(lambda x: a if ((x['director'] ==i) & 
                                             (pd.isna(x['country'])==True)) else x['country'], axis=1)

In [375]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      8807 non-null   object
 4   cast          8807 non-null   object
 5   country       8807 non-null   object
 6   date_added    8807 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8807 non-null   object
 9   duration      8807 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [376]:
df

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water           unknow   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans           unknow   
4         s5  TV Show           Kota Factory           unknow   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb           unknow   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                unknow  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...         unknow   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...  United States   
3                                                unknow         unknow   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...         unknow   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                             unknow         unknow   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

In [347]:
# a = df[(df['director']=='Mike Flanagan') & (df['country']!='N')]['country'].iloc[0]
a = df[(df['director']=='Mike Flanagan') & (df['country'].isnull())==False]['country'].iloc[0]
a

'United States'

In [348]:
pd.isna(df[df['director']=='Mike Flanagan']['country'].iloc[0]) 

True

In [349]:
# df['country'] = df.apply(lambda x: a if ((x['director'] =='Mike Flanagan') & (x['country'] =='N')) else x['country'], axis=1)
df['country'] = df.apply(lambda x: a if ((x['director'] =='Mike Flanagan') & 
                                         (pd.isna(x['country'])==True)) else x['country'], axis=1)

In [351]:
df

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water           unknow   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans           unknow   
4         s5  TV Show           Kota Factory           unknow   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb           unknow   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                unknow  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...         unknow   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                unknow         unknow   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...         unknow   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                             unknow         unknow   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]